src/rust/engine/src/externs/fs.rs

```diff
@@ -199,6 +207,68 @@ impl PySnapshot {
   }
 }
 
+#[pyclass]
+#[derive(Clone, Debug, PartialEq)]
+pub struct PyAddPrefix(pub Digest, pub PathBuf);
+
+#[pymethods]
+impl PyAddPrefix {
+  #[new]
+  fn __new__(digest: PyDigest, prefix: PathBuf) -> Self {
+    Self(digest.0, prefix)
+  }
+
+  fn __hash__(&self) -> u64 {
+    let mut s = DefaultHasher::new();
+    self.0.hash.prefix_hash().hash(&mut s);
+    self.1.hash(&mut s);
+    s.finish()
+  }
+
+  fn __repr__(&self) -> String {
+    format!("AddPrefix('{}', {})", PyDigest(self.0), self.1.display())
```

Comment:
Shouldn't `.display` be called automatically for `self.1` when used with the `{}` format specifier?

Reply:
No: `Path` doesn't implement `Display` because of the non-unicode paths issue, iirc.

Reply:
I think keeping as `PathBuf` still makes sense. This is a path. 

In fact, I would rather we could store directly as `RelativePath` - I didn't try implementing `ToPyObject` etc on it yet though, and I'm not sure it would work with the orphan rule.

Reply:
> In fact, I would rather we could store directly as RelativePath - I didn't try implementing ToPyObject etc on it yet though, and I'm not sure it would work with the orphan rule.

Either implement `ToPyObject` in the same crate where `RelativePath` is defined or just implement it for the `PyAddPrefix` / `PyRemoveStructs` structs.